### ISSUE TITLE
Equirectangular poses oriented in all four quadrants

### ISSUE THREAD

**Issue by sadams207** [openMVG/openMVG#1960]
Thank you for all the work on version 2.0. I just recently upgraded in order to hopefully solve a problem that I have been seeing with the poses computed from incremental SFM on equirectangular images. Attached is an image I generated to demonstrate the issue. Each sphere is a mesh textured with the equirectangular image. Axes are overlaid on each representing the local coordinate system. As I walked around the rectangular region of a building, the camera was always facing forward along the direction of travel. I expected the local coordinates to line up along the direction of travel, but they flip sign on the short sides of the rectangle. In other words, I expected the red lines to point opposite the direction of travel. The results seem consistent in quadrants 1 and 3 and seem to flip in quadrants 2 and 4, but I have not studied the problem enough to say definitively. To get to this result, I wrote out sfm_data to a json file, put the rotation and translations into a pose matrix, and multiplied by another pose to put the results in the xy plane. No trig or if statements involved, just matrix multiplication. I am not saying that I definitely did not create the issue. I'll try to setup a minimal example with four images, but any guidance or questions would be appreciated in the mean time. 
![image](https://user-images.githubusercontent.com/34065308/140545962-dc100d79-b9f8-40c2-99e3-6b00d5784f50.png)

**Comment by pmoulon** [comment]
From what I see the red Axis are consistent with your imagery (the red lines going to the back of your helmet (opposite side of the visible lens)).
Can you describe how do you plot those axes... Are you using the rotation matrix like rot * (0,0,1) for example... or the transpose?

1. Does the point cloud look consistent
2. Does the camera path seems consistent
3. Find the right way to fix the camera poses to fit your local coordinate system

**Comment by sadams207** [comment]
I agree with your comment - the red axis is consistent with the imagery. The spheres are textured in a local coordinate system and then oriented using pose/red axis. What is odd is that the ones on the short side appear rotated 180deg from reality.  

To plot the axes I am using the columns of the rotation matrix. 

The point cloud does look consistent on all four sides of the path. There were some objects along the halls and they were well defined and on the correct sides.

The camera path also seems consistent. It is smooth, continuous, and follows the expected path. 

I'll try to verify that if I rotate the poses on the ends of the rectangle that everything looks correct then. I'll also try to make a minimal example to make it easier to share. There were about 500 images used to construct this scene. I down sampled based on location to get what is shown.

**Comment by pmoulon** [comment]
Feel free to experiment with those small dataset too https://github.com/openMVG/Image_datasets/tree/main/360_dataset

**Comment by sadams207** [comment]
My bad. When I used the rotation matrix, to transform from local coordinates to world coordinates, apparently I should have been using the inverse. When I combined the rotation matrix with the center to form a pose transformation matrix, the two parts were incompatible. By simply using the inverse of the rotation matrix with the centers, it seems like all of the "issues" go away. 

Thank you for taking the time to address my questions.
